### PR TITLE
ref: Use Rate Limits

### PR DESCRIPTION
**Pull request by mgaeta**
## Rate Limits 
Copying the pattern from `baton/confluence-datacenter` where we try to forward along rate limit data instead of calling `wait()` to stall out for 1 second.

## Better Pagination
Originally, `Users.List()` fetched every group, stored them in memory, and then found all of the group memberships for each group. We have to do this convoluted API dance because Confluence doesn't expose an endpoint that lists all users. One benefit of keeping everything in memory  was that we can eliminate duplicate users.

~~This refactored method uses some non-deprecated endpoints to achieve the same thing, except that it uses the pagination bag pattern to break up the API calls.~~

Now we use the Search API to achieve the same thing. We might run into problems if a customer has more than 10,000 users, as the API docs mention that search queries return "up to 10k users".

The tests from the previous PR should still pass, expect we do not dedupe users.

## Summary by CodeRabbit

- **New Features**
	- Introduced methods for better user and group management, including options to add and remove users from groups.
	- Enhanced URL construction with a flexible mechanism for adding query parameters and pagination support.
	- Added a new helper function for improved pagination handling in user retrieval.

- **Bug Fixes**
	- Improved error handling related to rate limits and API responses.

- **Refactor**
	- Streamlined user and group management methods for clarity and efficiency.
	- Restructured URL generation functions to reduce errors and enhance clarity.
	- Refactored pagination handling for user retrieval to improve maintainability.

- **Tests**
	- Updated tests to validate new pagination logic and resource retrieval expectations.

### CODE REVIEW

File: pkg/connector/helpers.go [hunk 1]
```diff
@@ -12,3 +12,14 @@ func annotationsForUserResourceType() annotations.Annotations {
 	annos.Update(&v2.SkipEntitlementsAndGrants{})
 	return annos
 }
+
+func WithRateLimitAnnotations(
```
Comment (ggreer) [inline pkg/connector/helpers.go:16]: A function like this one should probably live in baton-sdk at some point.

File: pkg/connector/user.go [hunk 2]
```diff
@@ -45,40 +46,185 @@ func userResource(ctx context.Context, user *client.ConfluenceUser) (*v2.Resourc
...
 
-		userCopy := user
-		ur, err := userResource(ctx, &userCopy)
+		for _, group := range groups {
+			logger.Debug(
+				"adding a group to the bag",
+				zap.String("id", group.Id),
+				zap.String("name", group.Name),
+			)
+			bag.Push(
```
Comment (ggreer) [inline pkg/connector/user.go:140]: I think this will exceed the max page token size of 4096 characters if the customer has a lot of groups.

**Review by ggreer — COMMENTED** [review]
I like the rate limits, but I'm not sure about listing groups to list users. I think we could get a similar performance benefit by adding http caching to baton-sdk and keeping the List() code simpler.

File: pkg/connector/user.go (hunk 2 above)
Comment (mgaeta) [reply pkg/connector/user.go:140]: Should I bring down the page size when getting the list of groups? Is there a way to make sure that there is only one page worth of groups in the bag at a time?

Comment (ggreer) [reply pkg/connector/user.go:140]: Hard to say. Also I'm pretty sure it's possible to have a user that isn't a member of any group.

**Comment by jirwin** [comment]
@coderabbitai review

File: pkg/connector/client/confluence.go [hunk 3]
```diff
@@ -341,3 +344,82 @@ func strToInt(s string) int {
...
+	pageSize int,
+) (
+	[]ConfluenceUser,
+	string,
+	*v2.RateLimitDescription,
+	error,
+) {
+	getUsersUrl, err := c.parse(
+		SearchUrlPath,
+		withLimitAndOffset(pageToken, pageSize),
```
Comment (ggreer) [inline pkg/connector/client/confluence.go:399]: This works once you add `withQueryParameters(map[string]interface{}{"cql": "type=user"}),`

File: pkg/connector/user.go [hunk 4]
```diff
@@ -45,20 +45,74 @@ func userResource(ctx context.Context, user *client.ConfluenceUser) (*v2.Resourc
...
+	page := b.PageToken()
+	size := pToken.Size
+	if size == 0 {
+		size = ResourcesPageSize
+	}
+	if page == "" {
+		page = "0"
+	}
+	return b, page, size, nil
+}
```
[the remark this answers is not in the document]
Comment (jirwin) [reply pkg/connector/user.go:84]: @coderabbitai please generate them

File: pkg/connector/helpers.go [hunk 5]
```diff
@@ -23,3 +28,17 @@ func WithRateLimitAnnotations(
...
 }
+
+// shouldIncludeUser only include extant, human users.
+func shouldIncludeUser(ctx context.Context, user client.ConfluenceUser) bool {
+	logger := ctxzap.Extract(ctx)
+	if user.AccountType != accountTypeAtlassian {
+		logger.Debug("confluence: user is not of type atlassian", zap.Any("user", user))
+		return false
+	}
+	if len(user.Operations) == 0 {
```
Comment (ggreer) [inline pkg/connector/helpers.go:39]: We're not expanding operations in the list users call, so we don't add any users.

GetUsersFromSearch() needs to add `"expand": "operations",` to its query params.

File: pkg/connector/user.go [hunk 6]
```diff
@@ -112,23 +112,57 @@ func (o *userResourceType) List(
...
+
+			outputResources = append(outputResources, newUserResource)
+		}
+
+		err = bag.Next(nextToken)
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		if bag.Current() == nil {
```
Comment (mgaeta) [inline pkg/connector/user.go:154]: @ggreer Is this how I should make sure both methods get run?

**Review by ggreer — APPROVED** [review]

**Pull request merged**